File: bindings/elixir/lib/proven/safe_string.ex
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

defmodule Proven.SafeString do
  @moduledoc """
  Safe string escaping operations via libproven FFI.

  All escaping is performed by the Idris 2 verified core, ensuring
  correctness of HTML, SQL, and JavaScript escaping.
  """

  @doc """
  Escape a string for safe HTML insertion.

  Returns `{:ok, escaped}` or `{:error, reason}`.
  """
  @spec escape_html(String.t()) :: {:ok, String.t()} | {:error, atom()}
  def escape_html(value) when is_binary(value) do
    Proven.NIF.nif_string_escape_html(value)
  end

  @doc """
  Escape a string for safe SQL interpolation.

  Note: Prefer parameterized queries over string interpolation.
  Returns `{:ok, escaped}` or `{:error, reason}`.
  """
  @spec escape_sql(String.t()) :: {:ok, String.t()} | {:error, atom()}
  def escape_sql(value) when is_binary(value) do
    Proven.NIF.nif_string_escape_sql(value)
  end

  @doc """
  Escape a string for safe JavaScript string literal insertion.

  Returns `{:ok, escaped}` or `{:error, reason}`.
  """
  @spec escape_js(String.t()) :: {:ok, String.t()} | {:error, atom()}
  def escape_js(value) when is_binary(value) do
    Proven.NIF.nif_string_escape_js(value)
  end

  @doc """
  Check if a binary is valid UTF-8.
  """
  @spec valid_utf8?(binary()) :: boolean()
  def valid_utf8?(data) when is_binary(data) do
    Proven.NIF.nif_string_is_valid_utf8(data)
  end
end

File: bindings/lua/proven/safe_string.lua
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe string operations for escaping and sanitization.
--- @module proven.safe_string

local safe_string = {}

--- Escape a string for safe HTML insertion.
--- @param value string
--- @return string
function safe_string.escape_html(value)
    return value
        :gsub("&", "&amp;")
        :gsub("<", "&lt;")
        :gsub(">", "&gt;")
        :gsub('"', "&quot;")
        :gsub("'", "&#x27;")
end

--- Escape a string for safe SQL interpolation.
--- Note: Prefer parameterized queries over string interpolation.
--- @param value string
--- @return string
function safe_string.escape_sql(value)
    return value:gsub("'", "''")
end

--- Escape a string for safe JavaScript string literal insertion.
--- @param value string
--- @return string
function safe_string.escape_js(value)
    return value
        :gsub("\\", "\\\\")
        :gsub('"', '\\"')
        :gsub("'", "\\'")
        :gsub("\n", "\\n")
        :gsub("\r", "\\r")
        :gsub("\t", "\\t")
end

--- Percent-encode a string for safe URL inclusion.
--- @param value string
--- @return string
function safe_string.escape_url(value)
    return value:gsub("([^%w%-%.%_%~])", function(c)
        return string.format("%%%02X", string.byte(c))
    end)
end

--- Safely truncate a string to a maximum length.
--- @param value string
--- @param max_length number
--- @param suffix string|nil (default "...")
--- @return string
function safe_string.truncate_safe(value, max_length, suffix)
    suffix = suffix or "..."

    if max_length < 0 then
        return ""
    end

    if #value <= max_length then
        return value
    end

    local suffix_len = #suffix
    if max_length <= suffix_len then
        return value:sub(1, max_length)
    end

    return value:sub(1, max_length - suffix_len) .. suffix
end

return safe_string
